### PR TITLE
[core] Fix the deadlock in submit task when actor failed.

### PR DESCRIPTION
**Pull request by fishbone**
Signed-off-by: Yi Cheng <chengyidna@gmail.com>

## Why are these changes needed?

When actor died, it'll send notification to core workers. Right now, sometimes, core worker will queue the task waiting for actor death info and pop it up for better usability. But in async cases, this is going to cause issues.

The callback might submit tasks which require holding the lock. But it's already being held. This is going to cause a deadlock.

This PR fixed this by moving the failure part out of the lock.
## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/core_worker/transport/direct_actor_task_submitter.cc [hunk 1]
```diff
@@ -286,19 +286,8 @@ void CoreWorkerDirectActorTaskSubmitter::DisconnectActor(
...
-        RAY_UNUSED(task_finisher_.MarkTaskReturnObjectsFailed(
-            net_err_task.second, error_type, &error_info));
-      }
-
-      // No need to clean up tasks that have been sent and are waiting for
-      // replies. They will be treated as failed once the connection dies.
-      // We retain the sequencing information so that we can properly fail
-      // any tasks submitted after the actor death.
+      // We need to execute this outside of the lock to prevent deadlock.
+      wait_for_death_info_tasks = std::move(queue->second.wait_for_death_info_tasks);
```
Comment (scv119) [inline src/ray/core_worker/transport/direct_actor_task_submitter.cc:290]: should we clear `queue->second.wait_for_death_info_tasks`? I'm under the impression the object being moved aways's state is undefined.

File: src/ray/core_worker/transport/direct_actor_task_submitter.cc [hunk 2]
```diff
@@ -307,6 +296,17 @@ void CoreWorkerDirectActorTaskSubmitter::DisconnectActor(
     }
   }
 
+  if (!wait_for_death_info_tasks.empty()) {
+    auto error_type = GenErrorTypeFromDeathCause(death_cause);
+    const auto error_info = GetErrorInfoFromActorDeathCause(death_cause);
+    RAY_LOG(INFO) << "Failing tasks waiting for death info, size="
+                  << wait_for_death_info_tasks.size() << ", actor_id=" << actor_id;
+    for (auto &net_err_task : wait_for_death_info_tasks) {
+      RAY_UNUSED(task_finisher_.MarkTaskReturnObjectsFailed(
```
Comment (scv119) [inline src/ray/core_worker/transport/direct_actor_task_submitter.cc:305]: so this will try to hold the lock again?

**Review by scv119 — APPROVED** [review]
this is very tricky bug! thanks for figuring it out.

File: src/ray/core_worker/transport/direct_actor_task_submitter.cc (hunk 2 above)
Comment (iycheng) [reply src/ray/core_worker/transport/direct_actor_task_submitter.cc:305]: Well, it won't, but the callback inside might. Check my test, I put `_on_completed` there and that function might.
It'll be set in async env, but i'm not sure how to construct a test case triggering this in async env.

Comment (iycheng) [reply src/ray/core_worker/transport/direct_actor_task_submitter.cc:307]: To be more specific, for this test case, it'll call submit task and submit task is trying to hold the lock. maybe we should use reentrant lock here.

**Comment by iycheng** [comment]
@scv119  the code changed a lot from your last review. There are more deadlocks...

To be honest, I'm not sure how to write test cases to cover all this. @simon-mo  maybe you have some ideas about this given that I only reproduce this with serve and crashing raylet.

**Comment by iycheng** [comment]
With this fixing, the testing looks good:
image


and the failures are returning 500

image


We should be able to tune serve deployment and timeout to achieve better result. I only have 25 users sending request, so the drop is expected (need to change to another way to send traffic).

**Pull request merged**